lib/cdo/high_frequency_reporter.rb

```diff
@@ -0,0 +1,76 @@
+require 'csv'
+require 'time'
+# Usage:
+#   errors = HighFrequencyReporter.new(Slack.new)
+#   errors.load!
+#   loop do
+#     <stuff>
+#     errors.record "Something went wrong"
+#   end
+#   errors.report!
+#   errors.save
+class HighFrequencyReporter
+  attr_accessor :new_events
+
+  def initialize(chat_client, channel, log_file_name)
+    @chat_client = chat_client
+    @channel = channel
+    # Load old events from disk
+    # [{name: "eventblah", reported_at: '2020010...'}, ...]
+    @old_events = []
+    @new_events = []
+    @filename = log_file_name
+  end
+
+  # Loads known events from previous runs from a file on disk
+  def load
+    # Format for log file: CSV
+    # @see https://ruby-doc.org/stdlib-2.6.1/libdoc/csv/rdoc/CSV.html
+    # Each row:
+    #   "error message","timestamp"
+    # Where error message is the full text of the error message (appropriately
+    #   escaped by the CSV library)
+    # And timestamp is the last time we reported this error to Slack,
+    #   recorded in ISO 8601 format
+    @old_events = CSV.read(@filename, headers: true).map do |row|
+      row.to_h.transform_keys(&:to_sym).tap do |h|
+        h[:reported_at] = Time.parse(h[:reported_at])
+      end
+    end
+  rescue
+    # Okay to do nothing here.
+  end
+
+  # Add a new event to our record for the current run
+  # @param [String] event_name
+  def record(event_name)
+    @new_events << {name: event_name, reported_at: Time.now}
+  end
+
+  # Reports alertable events to chat client
+  # if current minute is multiple of throttle.
+  # Defaults to always report alertable events unless throttled.
+  # @param [Integer] throttle
+  def report!(throttle = 1)
+    if Time.now.min % throttle == 0
+      alertable_events.each {|e| @chat_client.message(e, {channel: @channel})}
+    end
+  end
+
+  # Saves known events and last report timestamps out to disk
+  def save
+    CSV.open(@filename, 'wb', headers: %i[name reported_at], write_headers: true) do |csv|
+      @new_events.each do |event|
+        csv << event.values
+      end
+    end
+  end
+
+  def alertable_events
+    @new_events.map {|e| e[:name]} & @old_events.map {|e| e[:name]}
```

Comment:
Does this `&` operator concatenate 2 arrays? I tried `[1,2] & [3,4]` and got `[]`.
Should we show the old events first to keep messages in chronological order?

Reply:
It's the intersection of the two -- we only want to report on events that have occurred in consecutive runs of the dropbox sync (most errors that have occurred so far have self-resolved, so we don't want to report those).